core/docs/features/sync-queue-design.md

```diff
@@ -0,0 +1,168 @@
+- `SYNC_DOWN` - stream is down. Contains stream id. Typical cause: remote node is down.
+- `SYNC_PONG` - pong response to the client's ping
+- `SYNC_CLOSE` - sync operation is closed (last response)
+
+## Stream Overview
+
+Stream consists of a sequence of numbered miniblocks and the recent events in the stream's minipool. 
+Each miniblock contains a sequence of events. Some miniblocks contain snapshots.
+
+Each stream is identified by unique stream id.
+
+Nodes host streams. Each node is identified by unique node address. If stream is hosted by the
+node, then stream is "local" to the node, otherwise it is "remote".
+
+Streams are replicated to a subset of nodes. If node is down, or there is other problem, sync can fail over to the next node.
+
+## Local and Remote Sync
+
+Node receiving `SyncStreams` from client streams updates for local streams directly to the client.
+
+For remote streams, node uses `SyncStreams` call to the remote nodes to request updates and forwards them to the client.
+
+If remote node becomes unavailable, node sends `SYNC_DOWN` to the client. Client retries with exponential backoff to
+add this stream again. Node should attempt to sync this stream from a different remote node if such request is received.
+
+## Implementation
+
+Propose design of the following components that interact with each other through asynchronous message passing (i.e. channels):
+
+- `SyncStreamHandler` and `SyncStreamHandlerRegistry` in package `syncv3/handler`
+- `EventBus` in package `syncv3/eventbus`
+- `LocalSyncer` in package `syncv3/local`
+- `RemoteSyncer` in package `syncv3/remote`
+- `SyncerRegistry` in package `syncv3/registry`
+
+All packages are subpackages of core/node/rpc/syncv3
+
+### `SyncStreamHandler` and `SyncStreamHandlerRegistry`
+
+`SyncStreamHandler` is responsible for handling single `SyncStreams` request, related `ModifySync` and `CancelSync` calls,
+and streaming updates to the client.
+
+`SyncStreamHandler` is created for each `SyncStreams` request. It is stored in `SyncStreamHandlerRegistry` and is removed
+when `SyncStreams` request is closed.
+
+`SyncStreamHandlerRegistry` is used to look up `SyncStreamHandler` by a sync id to dispatch `ModifySync` and `CancelSync` calls.
+
+`SyncStreamHandler` sends messages to `EventBus` to subscribe or to unsubscribe from the streams.
+
+`EventBus` streams all potential updates for the requested streams to the `SyncStreamHandler` queue. When `Subscribe` message is received,
+`EventBus` requests backfill for the given stream from the `LocalSyncer` or `RemoteSyncer`. `SyncStreamHandler` can receive updates
+before relevant backfill is complete and should filter such updates out and not send them to the client.
+
+`SyncStreamHandler` and `EventBus` communicate only through message passing. `SyncStreamHandler` provides `OnHandlerUpdate` interface which
+provides relevant update methods, implementations of these methods save updates to the `SyncStreamHandler` internal queue.
```

Comment:
> `SyncStreamHandler` provides `OnHandlerUpdate` interface which provides relevant update methods

To be clear, there will be just one `OnHandlerUpdate` function? I'm a bit confused by the `relevant update methodS`? Want to make sure that we keep ordering of updates by design and not by implementation.